### ISSUE TITLE
Heisenbug: use locale; ... (substr(...) cmp substr(...))

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3982]
Migrated from [rt.perl.org#6985](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=6985) (status was 'resolved')

Searchable as RT6985$

**Comment by p5pRT** [comment]
### From sburke@spinn.net

The bug seems unaffected by the presence or absensce of "use utf8"\.


I've observed this bug under 5\.6\.0 under MSWin \(ME and 98\) /and/ Linux
\(three different machines incidentally\)\, and under 5\.00305 under
AIX; those are all the systems I've tested this under\.

Below are perl \-V outputs on the three platforms\, and then a
program reproducing the bug in many iterations\.

I note that the bug seems to appear /only/ when there is an OS locale
active /and/ use locale is on /and/ we try to evaluate substr\(\.\.\.\)
cmp substr\(\.\.\.\) in scope of the "use locale"\.
Very odd indeed\!  Good luck finding a fix for this\.


Summary of my perl5 \(5\.0 patchlevel 5 subversion 3\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=aix\, osvers=4\.3\.1\.0\, archname=aix
&nbsp;    uname='aix pegasus 3 4 000165185700 '
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef useperlio=undef d\_sfio=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, optimize='\-O'\, gccversion=
&nbsp;    cppflags='\-D\_ALL\_SOURCE \-D\_ANSI\_C\_SOURCE \-D\_POSIX\_SOURCE \-qmaxmem=8192 \-I/usr/local/include \-I/usr/local/gnu/include'
&nbsp;    ccflags ='\-D\_ALL\_SOURCE \-D\_ANSI\_C\_SOURCE \-D\_POSIX\_SOURCE \-qmaxmem=8192 \-I/usr/local/include \-I/usr/local/gnu/include'
&nbsp;    stdchar='unsigned char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=8
&nbsp;    alignbytes=8\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='ld'\, ldflags ='\-L/usr/local/lib \-L/usr/local/gnu/lib \-L/usr/ccs/lib \-L/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib /usr/ccs/lib /usr/local/gnu/lib
&nbsp;    libs=\-lnsl \-lgdbm \-ldbm \-ldl \-lld \-lm \-lc \-lcrypt \-lbsd \-lPW
&nbsp;    libc=\, so=a\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_aix\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-bE&#8203;:perl\.exp'
&nbsp;    cccdlflags=' '\, lddlflags='\-bhalt&#8203;:4 \-bM&#8203;:SRE \-bI&#8203;:$\(PERL\_INC\)/perl\.exp \-bE&#8203;:$\(BASEEXT\)\.exp \-b noentry \-lc \-L/usr/local/lib \-L/usr/local/gnu/lib \-L/usr/ccs/lib \-L/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Built under aix
&nbsp;  Compiled at Aug 12 1999 09&#8203;:10&#8203;:29
&nbsp;  %ENV&#8203;:
&nbsp;    PERLLIB="/nfs/user/l/lachler/\.bin/perl/aix&#8203;:/nfs/user/l/lachler/\.bin/perl"
\[and\, not shown\, LANG="en\_US"\]

&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /nfs/user/l/lachler/\.bin/perl/aix
&nbsp;    /nfs/user/l/lachler/\.bin/perl
&nbsp;    /usr/local/gnu/lib/perl5/5\.00503/aix
&nbsp;    /usr/local/gnu/lib/perl5/5\.00503
&nbsp;    /usr/local/gnu/lib/perl5/site\_perl/5\.005/aix
&nbsp;    /usr/local/gnu/lib/perl5/site\_perl/5\.005
&nbsp;    \.




Summary of my perl5 \(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.2\.16\-21\, archname=i686\-linux
&nbsp;    uname='linux spinntwo 2\.2\.16\-21 \#1 wed aug 9 10&#8203;:38&#8203;:45 edt 2000 i686 unknown '
&nbsp;    config\_args=''
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=define 
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, optimize='\-O2'\, gccversion=egcs\-2\.91\.66 19990314/Linux \(egcs\-1\.1\.2 release\)
&nbsp;    cppflags='\-fno\-strict\-aliasing \-I/usr/local/include'
&nbsp;    ccflags ='\-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
&nbsp;    stdchar='char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lndbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt
&nbsp;    libc=/lib/libc\-2\.1\.3\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: USE\_LARGE\_FILES
&nbsp;  Built under linux
&nbsp;  Compiled at Oct 27 2000 18&#8203;:12&#8203;:05
&nbsp;  %ENV&#8203;:
&nbsp;    PERLLIB="/home/users/s/sburke/\.bin/perl/"
\[and\, not shown\, LANG="en\_US"\]

&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/users/s/sburke/\.bin/perl/
&nbsp;    /usr/lib/perl5/5\.6\.0/i686\-linux
&nbsp;    /usr/lib/perl5/5\.6\.0
&nbsp;    /usr/lib/perl5/site\_perl/5\.6\.0/i686\-linux
&nbsp;    /usr/lib/perl5/site\_perl/5\.6\.0
&nbsp;    /usr/lib/perl5/site\_perl/5\.005
&nbsp;    /usr/lib/perl5/site\_perl
&nbsp;    \.



Summary of my perl5 \(revision 5 version 6 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=MSWin32\, osvers=4\.0\, archname=MSWin32\-x86\-multi\-thread
&nbsp;    uname=''
&nbsp;    config\_args='undef'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=undef
&nbsp;    usethreads=undef use5005threads=undef useithreads=define usemultiplicity=define
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=undef 
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cl'\, optimize='\-O1 \-MD \-DNDEBUG'\, gccversion=
&nbsp;    cppflags='\-DWIN32'
&nbsp;    ccflags ='\-O1 \-MD \-DNDEBUG \-DWIN32 \-D\_CONSOLE \-DNO\_STRICT \-DHAVE\_DES\_FCRYPT  \-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS \-DPERL\_MSVCRT\_READFIX'
&nbsp;    stdchar='char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=10
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=4
&nbsp;    alignbytes=8\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='link'\, ldflags ='\-nologo \-nodefaultlib \-release  \-libpath&#8203;:"C&#8203;:\\Perl\\lib\\CORE"  \-machine&#8203;:x86'
&nbsp;    libpth="C&#8203;:\\Perl\\lib\\CORE"
&nbsp;    libs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib  comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib oleaut32\.lib  netapi32\.lib uuid\.lib wsock32\.lib mpr\.lib winmm\.lib  version\.lib odbc32\.lib odbccp32\.lib msvcrt\.lib
&nbsp;    libc=msvcrt\.lib\, so=dll\, useshrplib=yes\, libperl=perl56\.lib
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags='\-dll \-nologo \-nodefaultlib \-release  \-libpath&#8203;:"C&#8203;:\\Perl\\lib\\CORE"  \-machine&#8203;:x86'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: MULTIPLICITY USE\_ITHREADS PERL\_IMPLICIT\_CONTEXT PERL\_IMPLICIT\_SYS
&nbsp;  Locally applied patches&#8203;:
&nbsp;      ActivePerl Build 623
&nbsp;  Built under MSWin32
&nbsp;  Compiled at Dec 15 2000 16&#8203;:27&#8203;:07
&nbsp;  %ENV&#8203;:
&nbsp;    PERLDOC\_PAGER="towrite"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    C&#8203;:/Perl/lib
&nbsp;    C&#8203;:/Perl/site/lib
&nbsp;    \.
\[and a Control Panel "Regional Settings" value of "English \(United States\)" \-\-
altho changing to different locales \(Russian\, French\, Italian\.\.\.\) doesn't
seem to make the bug go away\.\]



And now the program exhibiting the bug&#8203;:

\# Under 5\.6\.0 and 5\.00503\, this screams like a baby under LANG=en\_US\,
\#  /or/ under MSWin\!
\# Commenting out 'use locale' makes the bug go away\, as does \(under
\#  UNIX\) unsetenv LANG\.
\# Time\-stamp&#8203;: "2001\-05\-10 01&#8203;:40&#8203;:16 MDT"

print "Perl version $\]\\n";
use strict;
use locale;
my $r = \.75;
my $bad = 0;
my $all = 0;
print "R&#8203;: $r\\n";
for\(1 \.\. 1000\) \{
&nbsp;  my $x = "foobar";
&nbsp;  my $y = "fooa";
&nbsp;  \($x\,$y\) = \($y\,$x\) if rand > $r;

&nbsp;  my $cmp1 = substr\($x\,0\,4\) cmp substr\($y\,0\,4\);
&nbsp;  my $i = substr\($x\,0\,4\);
&nbsp;  my $j = substr\($y\,0\,4\);
&nbsp;  my $cmp2 = $i cmp $j;

&nbsp;  \# The following two lines are never run&#8203;:
&nbsp;  print "\*\*\*\*SCREAM1\\n" unless $i eq substr\($x\,0\,4\);
&nbsp;  print "\*\*\*\*SCREAM2\\n" unless $j eq substr\($y\,0\,4\);
&nbsp;  
&nbsp;  if\($cmp1 \!= $cmp2\) \{
&nbsp;   \+\+$bad;
&nbsp;   print "SCREAM\!  ";
&nbsp;  \} else \{
&nbsp;   print  "okay     ";
&nbsp;  \}
&nbsp;  \+\+$all;
&nbsp;  
&nbsp;  printf "%2s : %2s \<%s>\<%s>  from \<%s> \<%s>\\n"\, 
&nbsp;    $cmp1\,$cmp2\,     $i\,$j\,     $x\,$y;
&nbsp;  
&nbsp;  
\}
printf "\\n\\nPct bad&#8203;: %g%% \(%g of %g\);  Cf\. r=%g\\n"\,
 100 \* $bad / $all\, $bad\, $all\,   $r;
\_\_END\_\_



Output&#8203;:

Perl version 5\.006
R&#8203;: 0\.75
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
\[\.\.\.lots more lines\.\.\.\]
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
okay     \-1 : \-1 \<fooa>\<foob>  from \<fooa> \<foobar>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>
SCREAM\!  \-1 :  1 \<foob>\<fooa>  from \<foobar> \<fooa>

Pct bad&#8203;: 72\.4% \(724 of 1000\);  Cf\. r=0\.75


You'll note that in the dump above\, you get a scream IFF "foob"
is first\.  Usually when I run the exact same program\, however\,
I get a scream IFF "foob" ISN'T first\.

As I run this more\, it seems to be the case that the pct\-bad
seems to vary directly with R /or/ 1\-R\.  With an R of \.75\, 
75% of the time\, pct\-bad is 25%\, and 25% of the time\, it's 75%\.
This relationship \(N% of runs report percent\-bads of approx 100\-N%\,
and vice versa\) seems to hold between \.5 and 1\, but seems to flip below
\.5 \(i\.e\.\, N% of runs report percent\-bads of approx N%\, and 100\-N%
of instances have 100\-N% percent\-bads\.\)

**Comment by p5pRT** [comment]
### From @jhi

&nbsp;    Fix for 20010514\.037; substr\(\) didn't invalidate the locale
&nbsp;    collation magic\.

Affected files \.\.\.

\.\.\. //depot/perl/pp\.c\#278 edit

Differences \.\.\.

==== //depot/perl/pp\.c\#278 \(text\) ====
Index&#8203;: perl/pp\.c
<details><summary>Inline Patch</summary>

```diff
--- perl/pp.c.~1~	Tue May 15 05:42:26 2001
+++ perl/pp.c	Tue May 15 05:42:26 2001
@@ -2808,6 +2808,9 @@
 	    sv_pos_u2b(sv, &pos, &rem);
 	tmps += pos;
 	sv_setpvn(TARG, tmps, rem);
+#ifdef USE_LOCALE_COLLATE
+	sv_unmagic(TARG, 'o');
+#endif
 	if (utf8_curlen)
 	    SvUTF8_on(TARG);
 	if (repl) {
End of Patch.
```
</details>
